File: mobile/lib/data/repositories/connection_repository.dart
import '../../input/dashboard_input.dart';
import '../../input/input_mapping.dart';
import '../../network/wheeldeck_client.dart';

/// Single source of truth for the WebSocket connection.
///
/// Thin pass-through over the [WheelDeckClient] service so ViewModels never
/// touch transport framing directly.
class ConnectionRepository {
  ConnectionRepository({required WheelDeckClient client}) : _client = client;

  final WheelDeckClient _client;

  ConnectionStatus get status => _client.status;
  ConnectionTarget? get lastTarget => _client.lastTarget;

  void onStatusChanged(void Function(ConnectionStatus status) callback) =>
      _client.onConnectionStatusChanged(callback);

  void onPairingRequired(void Function(PairingChallenge challenge) callback) =>
      _client.onPairingRequired(callback);

  void onPairingAccepted(void Function(String sessionToken) callback) =>
      _client.onPairingAccepted(callback);

  void setSessionToken(String? token) => _client.setSessionToken(token);

  Future<void> connect(ConnectionTarget target) => _client.connect(target);

  Future<void> disconnect() => _client.disconnect();

  /// Raw client for screens that still bind to it directly (settings,
  /// driving view during migration). New code should use this repository.
  WheelDeckClient get client => _client;

  /// Sends a `state` frame with the current input values.
  void sendState({
    required double steering,
    required double accelerator,
    required double brake,
    required double clutch,
  }) =>
      _client.sendState(
        steering: steering,
        accelerator: accelerator,
        brake: brake,
        clutch: clutch,
      );

  /// Sends a `button` frame for a dashboard control event.
  void sendButtonEvent(ControlId control, ActionType action) =>
      _client.sendButtonEvent(control, action);

  /// Sends the desired dashboard input mapping (`keyboard` or `gamepad`).
  void sendMappingMode(InputMapping mapping) =>
      _client.sendMappingMode(mapping);
}

File: android/app/src/main/kotlin/dev/fazrigading/wheeldeck/data/repositories/ConnectionRepository.kt
package dev.fazrigading.wheeldeck.data.repositories

import dev.fazrigading.wheeldeck.data.services.WheelDeckClient
import dev.fazrigading.wheeldeck.domain.models.ConnectionStatus
import dev.fazrigading.wheeldeck.domain.models.ConnectionTarget
import kotlinx.coroutines.flow.StateFlow

/// Single source of truth for the WebSocket connection.
///
/// Thin pass-through over the [WheelDeckClient] service so ViewModels never
/// touch transport framing directly.
class ConnectionRepository(private val client: WheelDeckClient) {
    val status: StateFlow<ConnectionStatus> = client.status
    val lastTarget: ConnectionTarget? get() = client.lastTarget

    fun connect(target: ConnectionTarget) = client.connect(target)

    fun disconnect() = client.disconnect()

    /// Sends a `state` frame with the current input values. The sequence
    /// number is the client's, so callers never build the wire type.
    fun sendState(
        steering: Double,
        accelerator: Double,
        brake: Double,
        clutch: Double,
        cameraX: Double = 0.0,
        cameraY: Double = 0.0,
    ) = client.sendState(
        steering = steering,
        accelerator = accelerator,
        brake = brake,
        clutch = clutch,
        cameraX = cameraX,
        cameraY = cameraY,
    )

    /// Sends a `button` frame for a dashboard control event.
    fun sendButtonEvent(control: String, action: String) = client.sendButtonEvent(control, action)

    /// Sends the desired dashboard input mapping (`keyboard` or `gamepad`).
    fun sendMappingMode(mode: String) = client.sendMappingMode(mode)

    /// Tells the desktop to revoke this pairing; send before disconnecting.
    fun sendUnpair() = client.sendUnpair()
}
